analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.Arithmetic.cs

```diff
@@ -55,58 +57,122 @@ private static NumberConstraint CalculateMultiply(NumberConstraint left, NumberC
+    private static BigInteger? CalculateAndMin(NumberConstraint left, NumberConstraint right)
     {
         // The result can only be negative if both operands are negative => The result must be >= 0 unless both ranges include negative numbers.
-        BigInteger? min = 0;
         if (left.CanBeNegative && right.CanBeNegative)
         {
-            min = left.Min.HasValue && right.Min.HasValue
-                ? NegativeMagnitude(BigInteger.Min(left.Min.Value, right.Min.Value))
-                : null;
+            return left.Min.HasValue && right.Min.HasValue ? NegativeMagnitude(BigInteger.Min(left.Min.Value, right.Min.Value)) : null;
+        }
+        else
+        {
+            return 0;
         }
-        return NumberConstraint.From(min, CalculateAndMax(left, right));
     }
 
     private static BigInteger? CalculateAndMax(NumberConstraint left, NumberConstraint right)
     {
         // BitAnd can only turn 1s into 0s, not the other way around => If both operands have the same sign, the result cannot be bigger than the smaller of the two.
-        if ((left.IsNegative && right.IsNegative)
-            || (left.IsPositive && right.IsPositive))
+        if ((left.IsNegative && right.IsNegative) || (left.IsPositive && right.IsPositive))
         {
             return SmallestMaximum(left, right);
         }
-        // -1 == 0b11111111 => a & -1 == a =>
-        // If one operand can be -1 and one is positive, the result will be positive, but not bigger than the positive operand.
         else if (left.IsPositive)
         {
+            // At this point, right range R can be:
+            // - negative => it can contain -1 or other negative numbers, which can at most make the result equal to max of L.
+            // - mixed => it's the union of a negative sub-range Rn and a positive one Rp.
+            //   - L & Rn can go as high as max of L, because Rn contains -1 (all ones).
+            //   - L & Rp, because both are positive, cannot be bigger than the smaller of max of L and Rp, so smaller of max of L and R.
+            // Therefore, L & R <= max(max of L, smaller of max of L and R) = max of L.
             return left.Max;
         }
         else if (right.IsPositive)
         {
+            // Same reasoning as above
             return right.Max;
         }
         else
         {
-            // If both operands can be -1 and both can have positive values => The result cannot be bigger than the bigger of the two maxima.
-            // If one operand is negative and one can have positive values => The result cannot be bigger than the positive maximum.
+            // If both ranges are mixed => They both contain -1 => The result can be as big as than the bigger of the two maxima.
+            // If one range is negative and one is mixed => The result can be as big as the max of the mixed range => bigger of the two maxima.
```

Comment:
We missed that we can simplify this. `BiggestMaximum` is just the general upper boundary—no need to explain it for specific cases.

Reply:
Indeed! Changed into `// If both ranges are mixed or one negative and one mixed => bigger of the two maxima is just the general upper boundary for max.`

Reply:
What about `// If both ranges are mixed or one negative and one mixed => use the general upper boundary for max.`

That the general upper boundary for max is the bigger of the two maxima is clear from the code.